zebra/zapi_msg.c

```diff
@@ -563,6 +563,9 @@ int zsend_redistribute_route(int cmd, struct zserv *client,
 
 	for (nexthop = re->nhe->nhg.nexthop;
 	     nexthop; nexthop = nexthop->next) {
+		if (CHECK_FLAG(nexthop->flags, NEXTHOP_FLAG_RECURSIVE))
```

Comment:
This is not quite correct as that nexthops are stored as ->next points to the next in the original chain of nexthops.  while ->resolved points to the resolved ones.  As I am reading this we are only sending the ->next nexthops.  We should be sending just the list of resolved nexthops up to the requestor.  This can be a combination of both ->next and ->resolved.  Imagine we have this

nh A
    resolved to B
    resolved to C
nh D
    resolved to E
    resolved to F

This code as I read it would just send B and C, We need to send B,C,E and F.

Reply:
Currently we send A and D without this change correct?

Reply:
## With and without the patch the RIB, OSPF database look the same

```
leaf2# show ip route 10.10.10.10/32 longer-prefixes 
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
       F - PBR, f - OpenFabric, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

B>  10.10.10.10/32 [200/0] via 192.168.1.0 (recursive), weight 1, 00:01:23
  *                          via 192.168.2.1, leaf2-eth0, weight 1, 00:01:23
                           via 192.168.1.2 (recursive), weight 1, 00:01:23
  *                          via 192.168.2.3, leaf2-eth1, weight 1, 00:01:23
leaf2# 

leaf2# show ip ospf database external 10.10.10.10

       OSPF Router with ID (192.168.2.0)

                AS External Link States 

  LS age: 140
  Options: 0x2  : *|-|-|-|-|-|E|-
  LS Flags: 0xb  
  LS Type: AS-external-LSA
  Link State ID: 10.10.10.10 (External Network Number)
  Advertising Router: 192.168.2.0
  LS Seq Number: 80000001
  Checksum: 0xc367
  Length: 36

  Network Mask: /32
        Metric Type: 2 (Larger than any link state path)
        TOS: 0
        Metric: 20
        Forward Address: 0.0.0.0
        External Route Tag: 0

leaf2#
```
The neighborship leaf2 < -- > exit1:
```
exit1# show ip route 10.10.10.10/32 longer-prefixes 
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
       F - PBR, f - OpenFabric, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

O>* 10.10.10.10/32 [110/20] via 192.168.3.1, exit1-eth0, weight 1, 00:01:56
exit1# 
```